### PR TITLE
ENH: Adds an out argument to bincount.

### LINKED ISSUE #9397: Feature request: Incremental bincount and axis argument
I have an application that requires to generate multiple histograms from a (25000,5000) array using bincount with 2^16 bins (16 bit values) each, e.g. the result would be a (65536,5000) array. The application first generates the initial histogram with apply_along_axis and then updates it up to 4000 times. Applying bincount along the axis is relatively fast, but adding the resulting (65536,5000) arrays is very slow. 

My application would significantly improve, if bincount would be able to update a previous array returned by bincount, e.g. if it could run incremental and if it could be applied along an axis similar to the sum or mean functions instead of using apply_along_axis.

### PR DESCRIPTION
**Pull request by jaimefrio**
Fixes #9397.

### CODE REVIEW

**Comment by jaimefrio** [comment]
It's missing all the necessary extra tests and additional documentation, and it uses `out=`, which doesn't seem to be a crowd favorite. But the implementation is all done and seems to be working fine on some informal testing.

File: numpy/core/src/multiarray/compiled_base.c [hunk 1]
```diff
@@ -57,66 +57,101 @@ check_array_monotonic(const double *a, npy_int lena)
...
+    static char *kwlist[] = {"list", "weights", "minlength", "out", NULL};
+
+    if (!PyArg_ParseTupleAndKeywords(args, kwds, "O|OOO:bincount", kwlist,
+                         &list_obj, &weights_obj, &minlength_obj, &out_obj)) {
+        goto fail;
     }
 
-    lst = (PyArrayObject *)PyArray_ContiguousFromAny(list, NPY_INTP, 1, 1);
-    if (lst == NULL) {
+    list_arr = _array_from_object(list_obj, NPY_INTP, 0);
```
Comment (jaimefrio) [inline numpy/core/src/multiarray/compiled_base.c]: Hmm... Perhaps we could mix in here some of the ideas in #7464, which I apparently abandoned for no good reason. Will give it some thought.

Comment (seberg) [inline numpy/core/src/multiarray/compiled_base.c]: I expect the biggest problem was what do about backcompat.

**Comment by charris** [comment]
Test failure looks legitimate.

**Comment by charris** [comment]
Couple of compiler warnings to deal with.

**Comment by mhvk** [comment]
This is still a nice idea. I'm not quite sure what got it stuck...

To get around the weirdness of not initializing `out`, one could also introduce an `initial` argument (and, then, if `initial is out`, one has no initialization).

File: numpy/core/src/multiarray/compiled_base.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (mhvk) [reply numpy/core/src/multiarray/compiled_base.c]: For this PR, isn't it fine to keep the *list* integer?

**Comment by GregoryMorse** [comment]
Please fix this PR as it would be a useful addition.

**Pull request closed without merging**